### ISSUE TITLE
Type checks: failure to infer type after match statement when returning from specific case statements

### ISSUE THREAD

**Issue by dermoth** [microsoft/pyright#7620]
## Environment data

-   Language Server version: 2024.4.1 (pyright version 1.1.356, commit 6652c4a8)
-   OS and version: Debian Bookworm (12.5)
-   Python version: OS-installed: python3 3.11.2-1+b1 (bookworm/main)

Relevant config:

- python.analysis.typeCheckingMode: standard

## Code Snippet

Pylance detects when types are being checked before use, so in the following snippet `f1()`and `f2()` have no type check errors, however in `f3()` it fails to detect the correct types as I used return instead of a default case. This is a common pattern using if statements as seen in `f1()`.

```python
from typing import Optional

def other(a: str, b: str) -> bool:
    return a == b

def f1(a: Optional[str], b: Optional[str]) -> bool:
    """f1"""

    if a is None and b is None:
        return True
    if a is None:
        return False
    if b is None:
        return False
    return other(a, b)

def f2(a: Optional[str], b: Optional[str]) -> bool:
    """f2"""

    match (a, b):
        case (None,None):
            return True
        case (_, None):
            return False
        case (None, _):
            return False
        case _:
            return other(a, b)

def f3(a: Optional[str], b: Optional[str]) -> bool:
    """f3"""

    match (a, b):
        case (None,None):
            return True
        case (_, None):
            return False
        case (None, _):
            return False
    return other(a, b)
```

In `f3()`, `return other(a, b)` is highlighted for both parameters with the following type check error:

```text
Argument of type "str | None" cannot be assigned to parameter "a" of type "str" in function "other"
  Type "str | None" cannot be assigned to type "str"
    "None" is incompatible with "str"Pylance[reportArgumentType](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportArgumentType)
-----
(parameter) a: str | None
```

## Expected behavior

The `f2()` and `f3()` functions are semantically identical. Pylance should ignore the types that end up in an unconditional return after checking in match/case statements, like it's doing for if statements.

## Logs

Let me know if additional logs are required.

**Comment by erictraut** [comment]
Thanks for reporting the issue. This will be addressed in the next release of pyright.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.358, which I just published.
